sticker/src/dataset.rs

```diff
@@ -0,0 +1,284 @@
+            .map(|s| s.pieces.len())
+            .max()
+            .unwrap_or(0);
+
+        let mut builder = TensorBuilder::new(
+            tokenized_sentences.len(),
+            max_seq_len,
+            self.encoders.iter().map(NamedEncoder::name),
+        );
+
+        for sentence in tokenized_sentences {
+            let input = self.vectorizer.vectorize(&sentence.pieces);
+            let mut token_mask = Array1::zeros((input.len(),));
+            for token_idx in &sentence.token_offsets {
+                token_mask[*token_idx] = 1;
+            }
+
+            let mut encoder_labels = HashMap::with_capacity(self.encoders.len());
+            for encoder in self.encoders {
+                let encoding = match encoder.encoder().encode(&sentence.sentence) {
+                    Ok(encoding) => encoding,
+                    Err(err) => return Some(Err(err)),
+                };
+
+                let mut labels = Array1::from_elem((input.len(),), 1i64);
+                for (encoding, offset) in encoding.into_iter().zip(&sentence.token_offsets) {
+                    labels[*offset] = encoding as i64;
+                }
+
+                encoder_labels.insert(encoder.name(), labels);
+            }
+
+            builder.add_with_labels(input.view(), encoder_labels, token_mask.view());
+        }
+
+        Some(Ok(builder.into()))
+    }
+}
+
+/// Trait providing adapters for `conllx::io::Sentences`.
+pub trait SentenceIter: Sized {
+    fn filter_by_len(self, max_len: usize) -> LengthFilter<Self>;
+    fn shuffle(self, buffer_size: usize) -> Shuffled<Self>;
+}
+
+impl<I> SentenceIter for I
+where
+    I: Iterator<Item = Fallible<Sentence>>,
+{
+    fn filter_by_len(self, max_len: usize) -> LengthFilter<Self> {
+        LengthFilter {
+            inner: self,
+            max_len,
+        }
+    }
+
+    fn shuffle(self, buffer_size: usize) -> Shuffled<Self> {
+        Shuffled {
+            inner: self,
+            buffer: RandomRemoveVec::with_capacity(buffer_size, XorShiftRng::from_entropy()),
+            buffer_size,
+        }
+    }
+}
+
+/// An Iterator adapter filtering sentences by maximum length.
+pub struct LengthFilter<I> {
+    inner: I,
+    max_len: usize,
+}
+
+impl<I> Iterator for LengthFilter<I>
+where
+    I: Iterator<Item = Fallible<Sentence>>,
+{
+    type Item = Fallible<Sentence>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(sent) = self.inner.next() {
+            // Treat Err as length 0 to keep our type as Result<Sentence, Error>. The iterator
+            // will properly return the Error at a later point.
+            let len = sent.as_ref().map(|s| s.len()).unwrap_or(0);
+            if len > self.max_len {
+                continue;
+            }
+            return Some(sent);
+        }
+        None
+    }
+}
+
+/// An Iterator adapter performing local shuffling.
+///
+/// Fills a buffer with size `buffer_size` on the first call. Subsequent
```

Comment:
Subsequent calls add the next incoming item to the buffer and pick a random element from the buffer.

Reply:
Updated.